stan/math/prim/mat/fun/gp_matern32_cov.hpp

```diff
@@ -0,0 +1,304 @@
+  for (size_t n = 0; n < x2_size; ++n)
+    check_not_nan(function, "x2", x2[n]);
+
+  check_positive_finite(function, "magnitude", sigma);
+  check_positive_finite(function, "length scale", length_scale);
+
+  T_s sigma_sq = square(sigma);
+  typename return_type<double, T_l>::type root_3_inv_l_sq
+      = sqrt(3.0) / length_scale;
+  typename return_type<double, T_l>::type neg_root_3_inv_l_sq
+      = -1.0 * sqrt(3.0) / length_scale;
+
+  for (size_t i = 0; i < x1_size; ++i) {
+    for (size_t j = 0; j < x2_size; ++j) {
+      typename return_type<T_x1, T_x2>::type distance
+          = sqrt(squared_distance(x1[i], x2[j]));
+      cov(i, j) = sigma_sq * (1.0 + root_3_inv_l_sq * distance)
+                  * exp(neg_root_3_inv_l_sq * distance);
+    }
+  }
+  return cov;
+}
+
+/**
+ * Returns a Matern 3/2 cross covariance matrix
+ *
+ * \f[ k(x, x') = \sigma^2(1 + \sqrt{3}
+ *   \sqrt{\sum_{k=1}^{K}\frac{d(x, x')^}{l_k^2}})
+ *   exp(-\sqrt{3}\sqrt{\sum_{k=1}^{K}\frac{d(x, x')^2}{l_k^2}}) \f]
+ *
+ * where \f$d(x, x')\f$ is the Euclidean distance
+ *
+ * This function is for the cross covariance matrix neededed
+ * to compute the posterior predictive density.
+ *
+ * @tparam T_x1 first type of std::vector of scalars
+ * @tparam T_x2 second type of std::vector of scalars
+ * @tparam T_s type of parameter sigma, marginal standard deviation
+ * @tparam T_l type of parameter length scale
+ *
+ * @param x1 std::vector of Eigen vectors of scalars
+ * @param x2 std::vector of Eigen vectors of scalars
+ * @param length_scale parameter length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x1, x2 are nan or inf
+ *
+ */
+template <typename T_x1, typename T_x2, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x1, T_x2, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern32_cov(const std::vector<Eigen::Matrix<T_x1, -1, 1>> &x1,
+                const std::vector<Eigen::Matrix<T_x2, -1, 1>> &x2,
+                const T_s &sigma, const std::vector<T_l> &length_scale) {
+  using std::exp;
+
+  size_t x1_size = size_of(x1);
+  size_t x2_size = size_of(x2);
+  Eigen::Matrix<typename return_type<T_x1, T_x2, T_s, T_l>::type,
+                Eigen::Dynamic, Eigen::Dynamic>
+      cov(x1_size, x2_size);
+
+  if (x1_size == 0 || x2_size == 0)
+    return cov;
+
+  const char *function = "gp_matern_32_cov";
+  for (size_t n = 0; n < x1_size; ++n)
+    check_not_nan(function, "x1", x1[n]);
+  for (size_t n = 0; n < x2_size; ++n)
+    check_not_nan(function, "x2", x2[n]);
+
+  check_positive_finite(function, "magnitude", sigma);
+  check_positive_finite(function, "length scale", length_scale);
+
+  size_t l_size = length_scale.size();
+  for (size_t n = 0; n < l_size; ++n)
+    check_not_nan(function, "length scale", length_scale[n]);
+
+  check_size_match(function, "x1 dimension", size_of(x1[0]),
```

Comment:
This function is inconsistent with the rest. The others make sure that all elements of `x1` are the same length and do not assume they are. Here, this only checks `x1[0]` and not `x1[1], ..., x1[x1_size-1]`.